exec/java-exec/src/main/java/org/apache/drill/exec/expr/stat/ParquetComparisonPredicate.java

```diff
@@ -83,37 +80,40 @@ private ParquetComparisonPredicate(
    * where Column1 and Column2 are from same parquet table.
    */
   @Override
-  public boolean canDrop(RangeExprEvaluator<C> evaluator) {
+  public RowsMatch matches(RangeExprEvaluator<C> evaluator) {
     Statistics<C> leftStat = left.accept(evaluator, null);
     if (isNullOrEmpty(leftStat)) {
-      return false;
+      return RowsMatch.SOME;
     }
-
     Statistics<C> rightStat = right.accept(evaluator, null);
     if (isNullOrEmpty(rightStat)) {
-      return false;
+      return RowsMatch.SOME;
     }
-
-    // if either side is ALL null, = is evaluated to UNKNOWN -> canDrop
     if (isAllNulls(leftStat, evaluator.getRowCount()) || isAllNulls(rightStat, evaluator.getRowCount())) {
-      return true;
+      return RowsMatch.NONE;
     }
+    return predicate.apply(leftStat, rightStat);
+  }
 
-    return (leftStat.hasNonNullValue() && rightStat.hasNonNullValue()) && predicate.test(leftStat, rightStat);
```

Comment:
Why check for `hasNonNullValue()` does not apply anymore?

Reply:
Condition tested inside checkNull when necessary

Reply:
I don't think it is the same check. Can you check if `isAllNulls` is equivalent to `!hasNonNullValue()` in which case consider replacing `isAllNulls()` on line 92.

Reply:
> Concerning the min/max not defined, here is an example I found in drill/exec/java-exec/src/test/resources/parquetFilterPush/blnTbl/0_0_3.parquet:
>
> value 1: R:0 D:0 V:
> value 2: R:0 D:0 V:
> value 3: R:0 D:0 V:
>
> row group 1: RC:3 TS:27 OFFSET:4
> col_bln: BOOLEAN UNCOMPRESSED DO:0 FPO:4 SZ:27/27/1.00 VC:3 ENC:BIT_PACKED,PLAIN,RLE ST:[num_nulls: 3, min/max not defined]

In this particular case `isAllNulls()` is `true` (all 3 values are `null`) and `!hasNonNullValue()` should be also `true` as there are no non `null` values.

Reply:
Debugger : 
System.out.println(leftStat) => min: false, max: false, num_nulls: 3
 System.out.println(leftStat.hasNonNullValue()) => true 
As you see, the function just states that min and max are defined, even if there are only nulls...

Reply:
Please provide details of the unit test.

Reply:
TestParquetFilterPushDown.testBooleanPredicate
"select col_bln from dfs.`parquetFilterPush/blnTbl` where not col_bln = true"
break point in ParquetComparisonPredicate.matches L.96
Print leftStat and leftStat.hasNonNullValue() when file = 0.0.3.parquet

Reply:
The statistics are incorrectly set by the drill, not by the parquet library. It is a bug in Metadata.java that uses `genericGetMin/Max()` without `hasNonNullValue()` check. I'll fix it in a separate PR.

Reply:
ok

Reply:
When `isAllNulls()` is `true`, `hasNonNullValue()` must be `false`, but if `isAllNulls()` is `false`, `hasNonNullValue()` may be `true` or `false` depending on data type and whether or not min and max can be calculated (not all data type have a proper comparator). It means that check for `hasNonNullValue()` can't be dropped (it is invalid to access or compare generic Min and/or Max without first checking that they are set).

Reply:
Yes I agree, but isn't it already checked in the previous function, matches?
public RowsMatch matches(RangeExprEvaluator<C> evaluator) {
if (isNullOrEmpty(leftStat)) {
      return RowsMatch.SOME;
    }
...
return predicate.apply(leftStat, rightStat);
}

where 
static boolean isNullOrEmpty(Statistics stat) {
    return stat == null || stat.isEmpty();
  }

where 
public boolean isEmpty() {
    return !this.hasNonNullValue && !this.isNumNullsSet();
  }

When I call predicate.apply, the check for hasNonNullValue is already done.

Reply:
It is possible that min and max are not set while the number of nulls is set, so `isEmpty()` is `false`.

Reply:
ok. hasNonNullValue() added in matches().